src/main/scala/scopt/options.scala

```diff
@@ -354,19 +354,24 @@ abstract case class OptionParser[C](programName: String) {
   }
   def optionsForRender: List[OptionDef[_, C]] = {
     val unsorted = options filter { o => o.kind != Head && o.kind != Check && !o.isHidden }
-    val (unseen, xs) = unsorted partition {_.hasParent} match {
+    val (remaining, sorted) = unsorted partition {_.hasParent} match {
       case (p, np) => (ListBuffer() ++ p, ListBuffer() ++ np)
     }
-    while (!unseen.isEmpty) {
+    var continue = true
```

Comment:
Kind of an ugly fix. I simply do a check to see if the `remaining` list has changed. If it's unchanged, then we quit to avoid infinite loop.

It sort of has to be fixed this way because a parent could be hidden, while children and especially grandchildren may not be hidden. So it would be insufficient for example to simply test to see if the direct parent was hidden.